platform-sdk/swirlds-common/src/main/java/com/swirlds/common/sequence/map/internal/AbstractSequenceMap.java

```diff
@@ -278,13 +294,18 @@ public boolean putIfAbsent(final K key, final V value) {
     @Override
     public V put(final K key, final V value) {
         final long sequenceNumber = getSequenceNumber(key);
-        final SequenceKeySet<K> keys = getSequenceKeySet(sequenceNumber);
+        SequenceKeySet<K> keys = getSequenceKeySet(sequenceNumber);
 
         lockSequenceNumber(sequenceNumber);
         try {
             if (keys.getSequenceNumber() != sequenceNumber) {
                 // the key is outside the allowed window
-                return null;
+                if (allowExpansion && sequenceNumber > getFirstSequenceNumberInWindow()) {
+                    expandCapacity(sequenceNumber);
```

Comment:
`fullyLock()` acquires each lock once, and `fullyUnlock()` releases each lock once.

```
    public void fullyLock() {
        for (int index = 0; index < parallelism; index++) {
            locks[index].lock();
        }
    }

    public void fullyUnlock() {
        for (int index = 0; index < parallelism; index++) {
            locks[index].unlock();
        }
    }

```

These are reentrant locks.

```
locks[lockIndex] = new ReentrantLock();
``

This means that a single thread is permitted to lock it multiple times without deadlocking itself. (i.e. it can re-enter a locked area of the code). The thread does not yield control of the lock until it has called unlock the same number of times that it has called lock.